### PR TITLE
k8s in custom Vnet - configurable kubeClusterCidr, kubeServiceCidr, kubeDnsServiceIp

### LINKED ISSUE #180: k8s: Make kubeClusterCidr, kubeServiceCidr, kubeDnsServiceIp configurable for custom VNet
In case you want to deploy a kubernetes cluster in an existing VNet which has a different address space than the default (10.0.0.0/8) you also want to adjust the hardcoded values for kubeClusterCidr, kubeServiceCidr, kubeDnsServiceIp. Currently the only way to achieve this is to deploy the templates and change the relevant files on the created master and agent vms. But you may still face problems.
It would be better to specify the values before template creation and get templates which fit the predefined VNet.

https://github.com/Azure/acs-engine/blob/2c4f3168751be5d3b2add8327eef8818c88daceb/parts/kubernetesmasteraddons-kube-dns-service.yaml#L11

https://github.com/Azure/acs-engine/blob/2c4f3168751be5d3b2add8327eef8818c88daceb/parts/kubernetesmastercustomdata.yml#L147

https://github.com/Azure/acs-engine/blob/2c4f3168751be5d3b2add8327eef8818c88daceb/parts/kubernetesmastervars.t#L58-L59

https://github.com/Azure/acs-engine/blob/2c4f3168751be5d3b2add8327eef8818c88daceb/pkg/acsengine/pki.go#L113

### PR DESCRIPTION
**Pull request by ashb**
This is a rebased version of #191.

The one change from the original pull request is where the config lives: rather than having a "kubeNetworkConfig" at the top level, I have addeded 3 keys to the existing orchestratorProfile.kubernetesConfig structure.

Closed #180, #191, #437

---
This change is [Reviewable](https://reviewable.io/reviews/azure/acs-engine/473)

### CODE REVIEW

File: pkg/api/converterfromapi.go [hunk 1]
```diff
@@ -285,6 +285,17 @@ func convertOrchestratorProfileToV20160330(api *OrchestratorProfile, o *v2016033
...
 	o.OrchestratorType = vlabs.OrchestratorType(api.OrchestratorType)
+
+	if api.KubernetesConfig != nil {
+		o.KubernetesConfig = &vlabs.KubernetesConfig{}
+		convertKubernetesConfigToVLabs(api.KubernetesConfig, o.KubernetesConfig)
+	}
+}
+
+func convertKubernetesConfigToVLabs(api *KubernetesConfig, vlabs *vlabs.KubernetesConfig) {
+	vlabs.DnsServiceIp = api.DnsServiceIP
```
Comment (JackQuincy) [inline pkg/api/converterfromapi.go]: seems like this is missing the inverse to these lines from convertertoapi.go

api.KubernetesImageBase = vlabs.KubernetesImageBase		  	
api.NetworkPolicy = vlabs.NetworkPolicy

could you add them please

File: pkg/acsengine/defaults.go [hunk 2]
```diff
@@ -209,6 +220,24 @@ func setDefaultCerts(a *api.Properties) (bool, error) {
...
+	_, serviceCidr, err := net.ParseCIDR(a.OrchestratorProfile.KubernetesConfig.ServiceCIDR)
+	if err != nil {
+		return false, err
+	}
+	// Based off https://play.golang.org/p/m8TNTtygK0
+	// Increment the last octet if it's zero.
+	firstIP := func(ip net.IP) net.IP {
+		for j := len(ip) - 1; j >= 0; j-- {
+			if ip[j] == 0 {
+				ip[j]++
```
Comment (JackQuincy) [inline pkg/acsengine/defaults.go]: Can you explain the need for this first ip? We didn't have this idea before. Was it covered from the first master ip?

**Comment by JackQuincy** [comment]
Also huge thanks for picking this work up.

**Comment by JackQuincy** [comment]
@acs-bot test this please

File: pkg/acsengine/defaults.go (hunk 2 above)
Comment (ashb) [inline pkg/acsengine/defaults.go]: I can now having played with it a bit more.

There is a service called `kubernetes` created in every cluster, and it has as it's IP the first IP in the service CIDR range. By default this is 10.0.0.1 (the default range is 10.0.0.0/16). Previously this was handled in CreatePki (and removed in this diff in the only changes in `pkg/acsengine/pki.go`), but rather than passing another argument in for it to then add to the IPs I handle it here.

The reason I do the ParseCIDR and then add the net is to be more lenient in the input - for instance this would cope with:

- 192.168.1.128/25 (192.168.1.128 would be the serviceIp)
- 192.168.1.1/25 or 192.168.1.0/25, both yielding 192.168.1.1 as the service IP etc.

**Comment by ashb** [comment]
I'll make the review feedback as requested in individual fixup! commits so it's hopefully easier to check the fixes, and then will squash to cleanliness before a final merge.

File: pkg/acsengine/const.go [hunk 3]
```diff
@@ -29,6 +29,15 @@ const (
 	DefaultInternalLbStaticIPOffset = 10
 	// DefaultNetworkPolicy is disabling network policy enforcement
 	DefaultNetworkPolicy = "none"
+	// DefaultKubernetsDnsServiceIP specifies the IP address that kube-dns
+	// listens on by default. must by in the default Service CIDR range.
+	DefaultKubernetsDnsServiceIP = "10.0.0.10"
```
Comment (colemickens) [inline pkg/acsengine/const.go]: There's a lot of `Kubernets` in here. (Missing the last `e`: `kubernets vs kubernetes`).

File: pkg/acsengine/defaults.go [hunk 4]
```diff
@@ -80,6 +80,17 @@ func setOrchestratorDefaults(cs *api.ContainerService) {
 		if a.OrchestratorProfile.KubernetesConfig.NetworkPolicy == "" {
 			a.OrchestratorProfile.KubernetesConfig.NetworkPolicy = DefaultNetworkPolicy
 		}
+
+		if a.OrchestratorProfile.KubernetesConfig.DnsServiceIP == "" {
+			a.OrchestratorProfile.KubernetesConfig.DnsServiceIP = DefaultKubernetsDnsServiceIP
+		}
+		if a.OrchestratorProfile.KubernetesConfig.ServiceCIDR == "" {
+			a.OrchestratorProfile.KubernetesConfig.ServiceCIDR = DefaultKubernetsServiceCIDR
```
Comment (colemickens) [inline pkg/acsengine/defaults.go]: Can you validate that the DnsServiceIP falls into the ServiceCIDR? (and isn't the .255 since I think you got tripped up on that)

Comment (ashb) [inline pkg/acsengine/defaults.go]: I can do that, and yes good call :)

Comment (ashb) [inline pkg/acsengine/defaults.go]: To make the validation a code nicer I'm going to mandate that dnsServiceIp and serviceCIDR have to both be specified if either is. 

(Otherwise the validation code would have to know about the default values - validation is currently done on the versioned API model, which hasn't yet had the defaults applied.)

**Comment by colemickens** [comment]
@acs-bot test this please

**Comment by ashb** [comment]
@colemickens @JackQuincy Thanks! All squashed to clean commits and conflicts resolved.

**Comment by JackQuincy** [comment]
@acs-bot test this please

File: pkg/acsengine/testdata/vnet/kubernetesvnet.json [hunk 5]
```diff
@@ -2,7 +2,12 @@
   "apiVersion": "vlabs",
   "properties": {
     "orchestratorProfile": {
-      "orchestratorType": "Kubernetes"
+      "orchestratorType": "Kubernetes",
+      "kubernetesConfig": {
+        "dnsServiceIp": "172.17.1.10",
```
Comment (amanohar) [inline pkg/acsengine/testdata/vnet/kubernetesvnet.json]: To make the naming convention consistent can you update this to: dnsService**IP** instead of dnsService**Ip** like in  firstConsecutiveStaticIP

Comment (ashb) [inline pkg/acsengine/testdata/vnet/kubernetesvnet.json]: Good call, have done.

**Comment by ashb** [comment]
@colemickens Any chance of a merge today?

**Comment by JackQuincy** [comment]
@acs-bot test this please

**Comment by JackQuincy** [comment]
We can merge once we pass ci/cd tests from @acs-bot

File: pkg/acsengine/pki.go [hunk 6]
```diff
@@ -117,8 +117,6 @@ func createCertificate(commonName string, caCertificate *x509.Certificate, caPri
 		template.KeyUsage |= x509.KeyUsageCertSign
 		template.IsCA = isCA
 	} else if isServer {
-		extraIPs = append(extraIPs, net.ParseIP("10.0.0.1"))
```
Comment (anhowe) [inline pkg/acsengine/pki.go]: Why was this removed?  Kubeproxy still writes an ip table rule, and kubedns relies on this.

Comment (anhowe) [inline pkg/acsengine/pki.go]: @ashb Why was this removed? Kubeproxy still writes an ip table rule, and kubedns relies on this.

Comment (colemickens) [inline pkg/acsengine/pki.go]: because it's done here: https://github.com/Azure/acs-engine/pull/473/files#diff-8edb2653431a5da9ad1f4286097702a7R224

Again, otherwise this wouldn't work in the first place.

Comment (anhowe) [inline pkg/acsengine/pki.go]: Thanks, I've read through the changes and they look good.

**Comment by lshahar** [comment]
I added to kubernetes.json example the following configuration under orchestratorProfile section
                "kubernetesConfig": {
                  "dnsServiceIP": "172.17.1.10",
                  "serviceCidr": "172.16.0.0/14",
                  "clusterCidr": "172.40.0.0/16"
 And try to generate the configuration by ./acs-engine example/kubernetes.json but under _output I still got configuration with 10.240.x, there is anymore configuration that should be done?

**Comment by ashb** [comment]
This will not change the network space the k8s creates. If you want to spin it up on a completely different subnet you will have to create the VNet and subnet yourself and use the "external VNet" example. Be careful that the subnet you choose doesn't overlap with the Docker subnet otherwise things might get.... interesting? (It might not, but the default docker subnet is 172.17.0.0/16 so....)

**Comment by colemickens** [comment]
To clarify,

@Ishahar, those options let you change:

* the ClusterIP used for KubeDNS. This must be determined ahead of time so that it can be configured on all `kubelet`s for the `--cluster-dns` flag.
* the service CIDR - a fake range used for ClusterIPs (the former must fall in this range)
* the cluster CIDR - the pod cidr used for the cluster

None of those change the actual IPs of the nodes. In that case, you'd want to do as @ashb suggests and use the existing vnet scenario examples (though now I'm wondering how that works, given that the PKI assets need to have the master IP known ahead of time... how does that even work @ashb ?)

**Comment by colemickens** [comment]
I guess in the existing vnet scenario, the assumption is still that k8s nodes can have free range on 10.x? I think this PR is not quite enough then, and we need to go further to allow flexibility of the node IP ranges as well.

Or is this PR enough to let me peer (or deploy to an existing vnet) with 10.x.x.x. stuff already in it, @ashb?

**Comment by ashb** [comment]
I'm just testing spinning up a k8s cluster in a VNet with an address space of 10.105.248/21, and I think, one bug with routetable association (debugging and will open an issue if so) it works and I can spin up a small cluster in a IP space fully under my control.

(The other downside right now is that the nodes are claiming a /24 for pods each where as I would really like it if I could make it /25 or /26. Again, different issue I think)

**Comment by ashb** [comment]
The PKI needs the master IP, but the master IP is predictable within the subnet namespace -- it is statically assigned based on `masterProfile.firstConsecutiveStaticIP` (which needs to be 14 IPs from the last IP in the subnet in practice for the internalLbIp to be valid). 

It's probably a bit error prone and must be configured exactly right, and I wouldn't want to gerenate the certs for it myself, but it is doable.
